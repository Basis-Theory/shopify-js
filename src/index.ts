export * from './EciesDecryptStrategy';
export * from './ShopifyPaymentSessionContext';
export * from './types';
