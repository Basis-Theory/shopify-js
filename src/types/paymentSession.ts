/* Shopify uses snake_case  */
/* eslint-disable camelcase */

interface PaymentMethod {
  type: string;
  data: {
    fingerprint: string;
    encrypted_message: string;
    ephemeral_public_key: string;
    tag: string;
  };
}

interface Address {
  given_name?: string;
  family_name: string;
  line1: string;
  line2?: string;
  city: string;
  postal_code?: string;
  province?: string;
  country_code: string;
  phone_number?: string;
  company?: string;
}

interface Customer {
  billing_address: Address;
  shipping_address: Address;
  email?: string;
  phone_number?: string;
  locale: string;
}

interface ClientDetails {
  ip_address: string;
  user_agent: string;
  accept_language: string;
}

interface PaymentSessionPayload {
  id: string;
  gid: string;
  group: string;
  amount: string;
  currency: string;
  test: boolean;
  merchant_locale: string;
  payment_method: PaymentMethod;
  proposed_at: string;
  customer?: Customer;
  kind: string;
  client_details?: ClientDetails;
}

/**
 * 2023-07 spec
 */
interface DecryptedPaymentMethod {
  type: string;
  data: {
    full_name: string;
    pan: string;
    month: number;
    year: number;
    verification_value: string;
  };
}

export type {
  PaymentSessionPayload,
  PaymentMethod,
  Address,
  Customer,
  ClientDetails,
  DecryptedPaymentMethod,
};
/* eslint-enable camelcase */
