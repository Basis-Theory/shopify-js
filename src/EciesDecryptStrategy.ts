import crypto from 'crypto';
import { DecryptedPaymentMethod, PaymentMethod } from './types';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const ECKey = require('ec-key');

interface DerivedKeyPair {
  cipherKey: string;
  hmacKey: string;
}

export class EciesDecryptStrategy {
  // There is no type definitions for ECKey
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly privateKey: any;

  public constructor(privateKey: Buffer | string) {
    this.privateKey = new ECKey(privateKey, 'pem');
  }

  /**
   * Decrypts a Shopify encrypted payment method.
   *
   * @param {PaymentMethod} paymentMethod - The encrypted payment method to decrypt.
   * @returns {DecryptedPaymentMethod} The decrypted payment method.
   * @throws {Error} If the decryption fails.
   */
  public decrypt(paymentMethod: PaymentMethod): DecryptedPaymentMethod {
    const {
      data: {
        ephemeral_public_key: ephemeralPublicKey,
        tag,
        encrypted_message: encryptedMessage,
      },
    } = paymentMethod;

    const sharedSecret = this.generateSharedSecret(ephemeralPublicKey);
    const { cipherKey, hmacKey } = this.deriveSymmetricKeyPair(sharedSecret);

    this.verifyMessageHmac(hmacKey, tag, encryptedMessage);

    const decrypted = this.decryptMessage(encryptedMessage, cipherKey);

    try {
      return JSON.parse(decrypted);
    } catch (error) {
      const err = new Error(
        'Unexpected format of decrypted data. Please check the encryption certificate fingerprint and private key.'
      );

      err.stack = error.stack;
      throw err;
    }
  }

  /**
   * Generates a shared secret for the provided ephemeral public key and the private key,
   * using ECDH (Elliptic Curve Diffie-Hellman)
   *
   * @private
   * @param {string} ephemeralPublicKey - The ephemeral public key in PEM format.
   * @returns {string} The shared secret in base64 encoding.
   */
  private generateSharedSecret(ephemeralPublicKey: string): string {
    const publicEc = new ECKey(ephemeralPublicKey, 'pem'); // Create a new ECKey instance from a base-64 spki string

    return this.privateKey.computeSecret(publicEc).toString('base64'); // Compute secret using private key for provided ephemeral public key
  }

  /**
   * Derives a symmetric key pair (cipher key and HMAC key) from the shared secret
   * using HKDF.
   *
   * @private
   * @param {string} sharedSecret - The shared secret in base64 encoding.
   * @returns {DerivedKeyPair} An object containing the cipher key and HMAC key.
   */
  private deriveSymmetricKeyPair(sharedSecret: string): DerivedKeyPair {
    const derived = Buffer.from(
      crypto.hkdfSync('sha256', Buffer.from(sharedSecret, 'base64'), '', '', 48)
    ).toString('hex');

    const cipherKey = derived.slice(0, 64);
    const hmacKey = derived.slice(-32);

    return {
      cipherKey,
      hmacKey,
    };
  }

  /**
   * Verifies the message authentication code (MAC) of the encrypted message.
   *
   * @private
   * @param {string} hmacKey - The HMAC key used to generate the hmac digest.
   * @param {string} tag - The expected HMAC tag in base64 encoding, computed separately by the sender.
   * @param {string} encryptedMessage - The encrypted message in base64 encoding.
   * @throws {Error} If the MAC verification fails.
   */
  private verifyMessageHmac(
    hmacKey: string,
    tag: string,
    encryptedMessage: string
  ): void {
    const hmac = crypto.createHmac('sha256', Buffer.from(hmacKey, 'hex'));

    hmac.update(Buffer.from(encryptedMessage, 'base64'));

    const digest = hmac.digest().toString('binary');
    const computedMac = digest.slice(0, 16);

    if (
      !crypto.timingSafeEqual(
        Buffer.from(computedMac, 'binary'),
        Buffer.from(tag, 'base64')
      )
    ) {
      throw new Error('Invalid Message Authentication Code');
    }
  }

  /**
   * Decrypts the encrypted message using the cipher key.
   *
   * @private
   * @param {string} encryptedMessage - The encrypted message in base64 encoding.
   * @param {string} cipherKey - The cipher key extracted from key derivation function.
   * @returns {string} The decrypted as UTF-8 string.
   */
  private decryptMessage(encryptedMessage: string, cipherKey: string): string {
    const decipher = crypto.createCipheriv(
      'aes-256-ctr',
      Buffer.from(cipherKey, 'hex'),
      Buffer.alloc(16)
    );

    return Buffer.concat([
      decipher.update(Buffer.from(encryptedMessage, 'base64')),
      decipher.final(),
    ]).toString('utf-8');
  }
}
