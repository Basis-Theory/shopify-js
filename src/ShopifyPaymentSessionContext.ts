import { EciesDecryptStrategy } from './EciesDecryptStrategy';
import { DecryptedPaymentMethod, PaymentMethod } from './types';

interface ShopifyPaymentSessionContextOptions {
  privateKeyPem: Buffer | string | { [fingerprint: string]: Buffer | string };
}

export class ShopifyPaymentSessionContext {
  public constructor(
    private readonly options: ShopifyPaymentSessionContextOptions
  ) {}

  public decrypt(paymentMethod: PaymentMethod): DecryptedPaymentMethod {
    let privateKey: Buffer | string;

    if (
      this.options.privateKeyPem instanceof Buffer ||
      typeof this.options.privateKeyPem === 'string'
    ) {
      privateKey = this.options.privateKeyPem;
    } else {
      const { fingerprint } = paymentMethod.data;

      privateKey = this.options.privateKeyPem[fingerprint];

      if (!privateKey) {
        throw new Error(
          `Couldn't locate private key for certificate fingerprint: ${fingerprint}`
        );
      }
    }

    const strategy = new EciesDecryptStrategy(privateKey);

    return strategy.decrypt(paymentMethod);
  }
}

export type { ShopifyPaymentSessionContextOptions };
