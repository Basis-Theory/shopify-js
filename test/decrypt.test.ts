import * as fs from 'fs';
import { ShopifyPaymentSessionContext } from '../src';
import paymentSession from './fixtures/paymentSession.json';

const privateKeyPem = fs.readFileSync(
  'test/fixtures/certificates/private_key.pem'
);
const wrongPrivateKeyPem = fs.readFileSync(
  'test/fixtures/certificates/wrong_private_key.pem'
);

describe('decrypt', () => {
  test('should throw when using wrong key', () => {
    const context = new ShopifyPaymentSessionContext({
      privateKeyPem: wrongPrivateKeyPem,
    });

    expect(() => context.decrypt(paymentSession.payment_method)).toThrow(
      'Invalid Message Authentication Code'
    );
  });

  test('should throw when using key not found', () => {
    const context = new ShopifyPaymentSessionContext({
      privateKeyPem: {
        '84d14a26c7d80e4975ed646f8ba8484a5ee8de970a04f1cf61b775aa7cbdc9cf':
          wrongPrivateKeyPem,
      },
    });

    expect(() => context.decrypt(paymentSession.payment_method)).toThrow(
      "Couldn't locate private key for certificate fingerprint: 23f3812e093737252e98d47aa790ba5e607188837ec9aea797c0dde8ed7ef674"
    );
  });

  test('should decrypt using right key', () => {
    const context = new ShopifyPaymentSessionContext({
      privateKeyPem: privateKeyPem.toString('utf-8'), // string or Buffer is acceptable
    });

    const decrypted = context.decrypt(paymentSession.payment_method);

    expect(decrypted).toMatchSnapshot();
  });

  test('should decrypt finding key by certificate fingerprint', () => {
    const context = new ShopifyPaymentSessionContext({
      privateKeyPem: {
        '23f3812e093737252e98d47aa790ba5e607188837ec9aea797c0dde8ed7ef674':
          privateKeyPem,
        '84d14a26c7d80e4975ed646f8ba8484a5ee8de970a04f1cf61b775aa7cbdc9cf':
          wrongPrivateKeyPem,
      },
    });

    const decrypted = context.decrypt(paymentSession.payment_method);

    expect(decrypted).toMatchSnapshot();
  });
});
