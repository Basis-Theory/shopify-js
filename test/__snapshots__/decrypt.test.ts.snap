// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decrypt should decrypt finding key by certificate fingerprint 1`] = `
{
  "data": {
    "full_name": "Jim S",
    "month": 4,
    "pan": "4242424242424242",
    "verification_value": "234",
    "year": 2024,
  },
  "type": "card",
}
`;

exports[`decrypt should decrypt using right key 1`] = `
{
  "data": {
    "full_name": "Jim S",
    "month": 4,
    "pan": "4242424242424242",
    "verification_value": "234",
    "year": 2024,
  },
  "type": "card",
}
`;
